File: sdk/typescript/src/namespaces/history.ts
/**
 * History Namespace API
 *
 * Provides access to historical data including debate history,
 * nomic cycles, and system events.
 */

import type { AragoraClient } from '../client';

/**
 * Historical debate entry
 */
export interface HistoricalDebate {
  id: string;
  question: string;
  started_at: string;
  ended_at?: string;
  status: 'completed' | 'active' | 'cancelled';
  winner?: string;
  consensus_reached: boolean;
  participants: string[];
  domain?: string;
  rounds: number;
}

/**
 * Nomic cycle entry
 */
export interface NomicCycle {
  cycle_id: string;
  phase: string;
  started_at: string;
  ended_at?: string;
  status: 'completed' | 'active' | 'failed';
  improvements_proposed: number;
  improvements_implemented: number;
  files_modified: string[];
  summary?: string;
}

/**
 * System event entry
 */
export interface HistoricalEvent {
  id: string;
  event_type: string;
  timestamp: string;
  actor?: string;
  resource_type?: string;
  resource_id?: string;
  details: Record<string, unknown>;
  severity: 'info' | 'warning' | 'error';
}

/**
 * History summary
 */
export interface HistorySummary {
  total_debates: number;
  debates_this_week: number;
  debates_this_month: number;
  total_nomic_cycles: number;
  active_agents: number;
  most_active_agent: string;
  top_domains: Array<{ domain: string; count: number }>;
  consensus_rate: number;
  average_debate_duration_minutes: number;
}

/**
 * History query options
 */
export interface HistoryQueryOptions {
  limit?: number;
  offset?: number;
  since?: string;
  until?: string;
  status?: string;
  agent?: string;
  domain?: string;
}

/**
 * History namespace for accessing historical data.
 *
 * @example
 * ```typescript
 * // Get recent debates
 * const debates = await client.history.listDebates({ limit: 10 });
 *
 * // Get history summary
 * const summary = await client.history.getSummary();
 * console.log(`Total debates: ${summary.total_debates}`);
 * ```
 */
export class HistoryNamespace {
  constructor(private client: AragoraClient) {}

  /**
   * List historical debates.
   *
   * @param options - Query options for filtering and pagination
   */
  async listDebates(options?: HistoryQueryOptions): Promise<HistoricalDebate[]> {
    const response = await this.client.request<{ debates: HistoricalDebate[] }>(
      'GET',
      '/api/v1/history/debates',
      { params: options as Record<string, unknown> }
    );
    return response.debates;
  }

  /**
   * List nomic improvement cycles.
   *
   * @param options - Query options for filtering and pagination
   */
  async listCycles(options?: HistoryQueryOptions): Promise<NomicCycle[]> {
    const response = await this.client.request<{ cycles: NomicCycle[] }>(
      'GET',
      '/api/v1/history/cycles',
      { params: options as Record<string, unknown> }
    );
    return response.cycles;
  }

  /**
   * List system events.
   *
   * @param options - Query options for filtering and pagination
   */
  async listEvents(options?: HistoryQueryOptions & {
    event_type?: string;
    severity?: 'info' | 'warning' | 'error';
  }): Promise<HistoricalEvent[]> {
    const response = await this.client.request<{ events: HistoricalEvent[] }>(
      'GET',
      '/api/v1/history/events',
      { params: options as Record<string, unknown> }
    );
    return response.events;
  }

  /**
   * Get a summary of historical activity.
   *
   * @param options.period - Time period for summary ('week', 'month', 'year', 'all')
   */
  async getSummary(options?: { period?: 'week' | 'month' | 'year' | 'all' }): Promise<HistorySummary> {
    return this.client.request<HistorySummary>('GET', '/api/v1/history/summary', {
      params: options,
    });
  }

  /**
   * Get history for a specific agent.
   *
   * @param agentName - The agent name
   * @param options - Query options
   */
  async getAgentHistory(
    agentName: string,
    options?: HistoryQueryOptions
  ): Promise<HistoricalDebate[]> {
    const response = await this.client.request<{ debates: HistoricalDebate[] }>(
      'GET',
      `/api/v1/agent/${encodeURIComponent(agentName)}/history`,
      { params: options as Record<string, unknown> }
    );
    return response.debates;
  }
}

File: sdk/python/aragora/namespaces/history.py
"""
History Namespace API

Provides access to historical data including debate history,
nomic cycles, and system events.

Features:
- List historical debates
- View nomic improvement cycles
- Track system events
- Get history summaries
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


HistoryPeriod = Literal["week", "month", "year", "all"]
EventSeverity = Literal["info", "warning", "error"]


class HistoryAPI:
    """
    Synchronous History API.

    Provides access to historical data:
    - List historical debates
    - View nomic improvement cycles
    - Track system events
    - Get history summaries

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> debates = client.history.list_debates(limit=10)
        >>> summary = client.history.get_summary()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list_debates(
        self,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
        status: str | None = None,
        agent: str | None = None,
        domain: str | None = None,
    ) -> list[dict[str, Any]]:
        """
        List historical debates.

        Args:
            limit: Maximum number of debates
            offset: Number of debates to skip
            since: Start date (ISO 8601)
            until: End date (ISO 8601)
            status: Filter by status
            agent: Filter by agent
            domain: Filter by domain

        Returns:
            List of historical debate records
        """
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        if status:
            params["status"] = status
        if agent:
            params["agent"] = agent
        if domain:
            params["domain"] = domain
        response = self._client.request(
            "GET", "/api/v1/history/debates", params=params if params else None
        )
        return response.get("debates", [])

    def list_cycles(
        self,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
        status: str | None = None,
    ) -> list[dict[str, Any]]:
        """
        List nomic improvement cycles.

        Args:
            limit: Maximum number of cycles
            offset: Number of cycles to skip
            since: Start date (ISO 8601)
            until: End date (ISO 8601)
            status: Filter by status

        Returns:
            List of nomic cycle records
        """
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        if status:
            params["status"] = status
        response = self._client.request(
            "GET", "/api/v1/history/cycles", params=params if params else None
        )
        return response.get("cycles", [])

    def list_events(
        self,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
        event_type: str | None = None,
        severity: EventSeverity | None = None,
    ) -> list[dict[str, Any]]:
        """
        List system events.

        Args:
            limit: Maximum number of events
            offset: Number of events to skip
            since: Start date (ISO 8601)
            until: End date (ISO 8601)
            event_type: Filter by event type
            severity: Filter by severity

        Returns:
            List of system event records
        """
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        if event_type:
            params["event_type"] = event_type
        if severity:
            params["severity"] = severity
        response = self._client.request(
            "GET", "/api/v1/history/events", params=params if params else None
        )
        return response.get("events", [])

    def get_summary(self, period: HistoryPeriod | None = None) -> dict[str, Any]:
        """
        Get a summary of historical activity.

        Args:
            period: Time period for summary

        Returns:
            Dict with:
            - total_debates: Total debate count
            - debates_this_week: Debates this week
            - debates_this_month: Debates this month
            - total_nomic_cycles: Nomic cycle count
            - active_agents: Active agent count
            - most_active_agent: Top agent
            - top_domains: Most debated domains
            - consensus_rate: Consensus success rate
            - average_debate_duration_minutes: Average duration
        """
        params = {"period": period} if period else None
        return self._client.request("GET", "/api/v1/history/summary", params=params)

    def get_agent_history(
        self,
        agent_name: str,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
    ) -> list[dict[str, Any]]:
        """
        Get history for a specific agent.

        Args:
            agent_name: The agent name
            limit: Maximum number of debates
            offset: Number to skip
            since: Start date
            until: End date

        Returns:
            List of debates involving the agent
        """
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        response = self._client.request(
            "GET",
            f"/api/v1/agent/{agent_name}/history",
            params=params if params else None,
        )
        return response.get("debates", [])


class AsyncHistoryAPI:
    """
    Asynchronous History API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     debates = await client.history.list_debates(limit=10)
        ...     summary = await client.history.get_summary()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list_debates(
        self,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
        status: str | None = None,
        agent: str | None = None,
        domain: str | None = None,
    ) -> list[dict[str, Any]]:
        """List historical debates."""
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        if status:
            params["status"] = status
        if agent:
            params["agent"] = agent
        if domain:
            params["domain"] = domain
        response = await self._client.request(
            "GET", "/api/v1/history/debates", params=params if params else None
        )
        return response.get("debates", [])

    async def list_cycles(
        self,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
        status: str | None = None,
    ) -> list[dict[str, Any]]:
        """List nomic improvement cycles."""
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        if status:
            params["status"] = status
        response = await self._client.request(
            "GET", "/api/v1/history/cycles", params=params if params else None
        )
        return response.get("cycles", [])

    async def list_events(
        self,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
        event_type: str | None = None,
        severity: EventSeverity | None = None,
    ) -> list[dict[str, Any]]:
        """List system events."""
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        if event_type:
            params["event_type"] = event_type
        if severity:
            params["severity"] = severity
        response = await self._client.request(
            "GET", "/api/v1/history/events", params=params if params else None
        )
        return response.get("events", [])

    async def get_summary(self, period: HistoryPeriod | None = None) -> dict[str, Any]:
        """Get a summary of historical activity."""
        params = {"period": period} if period else None
        return await self._client.request("GET", "/api/v1/history/summary", params=params)

    async def get_agent_history(
        self,
        agent_name: str,
        limit: int | None = None,
        offset: int | None = None,
        since: str | None = None,
        until: str | None = None,
    ) -> list[dict[str, Any]]:
        """Get history for a specific agent."""
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if since:
            params["since"] = since
        if until:
            params["until"] = until
        response = await self._client.request(
            "GET",
            f"/api/v1/agent/{agent_name}/history",
            params=params if params else None,
        )
        return response.get("debates", [])
